cmd/atlasci/internal/ci/ci.go

```diff
@@ -0,0 +1,176 @@
+	"github.com/go-git/go-git/v5/utils/merkletrie"
+)
+
+type (
+	// A ChangeDetector takes a migration directory and splits it into the "base" files (already merged) and new ones.
+	ChangeDetector interface {
+		// DetectChanges splits the files of a migration directory into the "base" files (already merged) and new ones.
+		DetectChanges(context.Context) ([]migrate.File, []migrate.File, error)
+	}
+	// A ChangeLoader takes a set of migration files and will create multiple schema.Changes out of it.
+	// It will also label migration files as either "generated" or "handcrafted".
+	ChangeLoader interface {
+		// LoadChanges converts each of the given migration files into one Changes.
+		LoadChanges(context.Context, []migrate.File) ([]Changes, error)
+	}
+	// DirScanner stitches migrate.Dir and migrate.Scanner into one interface.
+	DirScanner interface {
+		migrate.Dir
+		migrate.Scanner
+	}
+	// Change wraps a schema.Change and adds metadata to it.
+	Change struct {
+		schema.Change
+		// Generated determines if this file was solely generated by Atlas and not edited later on by a human.
+		Generated bool
+	}
+	// Changes is a list of Change.
+	Changes []Change
+)
+
+type (
+	// GitChangeDetector implements the ChangeDetector interface by utilizing a git repository.
+	GitChangeDetector struct {
+		root string     // path to the git repository root
+		base string     // name of the base branch (e.g. master)
+		path string     // path of the migration directory relative to the repository root (in slash notation)
+		dir  DirScanner // the migration directory to load migration files from
+	}
+	// GitChangeDetectorOption allows configuring GitChangeDetector with functional arguments.
+	GitChangeDetectorOption func(*GitChangeDetector) error
+)
+
+// NewGitChangeDetector configures a new GitChangeDetector.
+func NewGitChangeDetector(root string, dir DirScanner, opts ...GitChangeDetectorOption) (*GitChangeDetector, error) {
+	if root == "" {
+		return nil, fmt.Errorf("internal/ci: root cannot be %q", root)
+	}
+	if dir == nil {
+		return nil, fmt.Errorf("internal/ci: dir cannot be %v", dir)
+	}
+	d := &GitChangeDetector{root: root, dir: dir}
+	for _, opt := range opts {
+		if err := opt(d); err != nil {
+			return nil, err
+		}
+	}
+	if d.base == "" {
+		d.base = "master"
+	}
+	if d.path == "" {
+		d.path = "migrations"
+	}
+	return d, nil
+}
+
+// WithBase configures the git base branch name for a GitChangeDetector.
+func WithBase(base string) GitChangeDetectorOption {
+	return func(d *GitChangeDetector) error {
+		d.base = base
+		return nil
+	}
+}
+
+// WithMigrationsPath configures the git base branch name for a GitChangeDetector.
+func WithMigrationsPath(path string) GitChangeDetectorOption {
+	return func(d *GitChangeDetector) error {
+		d.path = filepath.ToSlash(path)
+		return nil
+	}
+}
+
+// DetectChanges implements the ChangeDetector interface.
+func (d *GitChangeDetector) DetectChanges(ctx context.Context) ([]migrate.File, []migrate.File, error) {
```

Comment:
I'm not sure the logic here is 100% correct, because the HEAD can squash the migration directory and delete old files. Thoughts?

Reply:
Currently it can't, since the sum.file prevent this (at least the executor will not execute this scenario, therefore it can be considered a user-error). This needs special handling, when Atlas supports squashing (or checkpointing).